@bors-servo: r-
```
error[E0382]: use of partially moved value: `fetch_response`

   --> /home/travis/build/servo/servo/tests/unit/net/fetch.rs:387:13

    |

386 |     assert!(fetch_response.url_list.into_inner().len() == 0);

    |             ----------------------- value moved here

387 |     assert!(fetch_response.url().is_none());

    |             ^^^^^^^^^^^^^^ value used here after move

    |

    = note: move occurs because `fetch_response.url_list` has type `std::cell::RefCell<std::vec::Vec<url::Url>>`, which does not implement the `Copy` trait

error: aborting due to previous error
```

--- a/components/net/fetch/methods.rs
+++ b/components/net/fetch/methods.rs
@@ -409,9 +409,7 @@ fn basic_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
 
     match url.scheme() {
         "about" if url.path() == "blank" => {
-            let mut response = Response::new();
-            // https://github.com/whatwg/fetch/issues/312
-            response.url = Some(url);
+            let mut response = Response::new(url);
             response.headers.set(ContentType(mime!(Text / Html; Charset = Utf8)));
             *response.body.lock().unwrap() = ResponseBody::Done(vec![]);
             response
@@ -425,9 +423,7 @@ fn basic_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
             if *request.method.borrow() == Method::Get {
                 match decode(&url) {
                     Ok((mime, bytes)) => {
-                        let mut response = Response::new();
-                        // https://github.com/whatwg/fetch/issues/312
-                        response.url = Some(url.clone());
+                        let mut response = Response::new(url);
                         *response.body.lock().unwrap() = ResponseBody::Done(bytes);
                         response.headers.set(ContentType(mime));
                         response
@@ -449,9 +445,7 @@ fn basic_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
                                 let _ = file.read_to_end(&mut bytes);
                                 let mime = guess_mime_type(file_path);
 
-                                let mut response = Response::new();
-                                // https://github.com/whatwg/fetch/issues/312
-                                response.url = Some(url.clone());
+                                let mut response = Response::new(url);
                                 *response.body.lock().unwrap() = ResponseBody::Done(bytes);
                                 response.headers.set(ContentType(mime));
                                 response
@@ -475,8 +469,7 @@ fn basic_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
 
             match load_blob_sync(url.clone(), context.filemanager.clone()) {
                 Ok((headers, bytes)) => {
-                    let mut response = Response::new();
-                    response.url = Some(url.clone());
+                    let mut response = Response::new(url);
                     response.headers = headers;
                     *response.body.lock().unwrap() = ResponseBody::Done(bytes);
                     response
@@ -1028,8 +1021,7 @@ fn http_network_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
         }
     };
 
-    let mut response = Response::new();
-    response.url = Some(url.clone());
+    let mut response = Response::new(url.clone());
     response.status = Some(res.response.status);
     response.raw_status = Some((res.response.status_raw().0,
                                 res.response.status_raw().1.as_bytes().to_vec()));

--- a/components/net/fetch/methods.rs
+++ b/components/net/fetch/methods.rs
@@ -687,7 +687,7 @@ fn http_redirect_fetch<UI: 'static + UIProvider>(request: Rc<Request>,
         Some(&Location(ref location)) => location.clone(),
         _ => return Response::network_error(NetworkError::Internal("Location header parsing failure".into()))
     };
-    let response_url = response.actual_response().url.as_ref().unwrap();
+    let response_url = response.actual_response().url().unwrap();
     let location_url = response_url.join(&*location);
     let location_url = match location_url {
         Ok(url) => url,

--- a/components/net_traits/response.rs
+++ b/components/net_traits/response.rs
@@ -79,7 +79,7 @@ pub enum ResponseMsg {
 pub struct Response {
     pub response_type: ResponseType,
     pub termination_reason: Option<TerminationReason>,
-    pub url: Option<Url>,
+    url: Option<Url>,
     pub url_list: RefCell<Vec<Url>>,
     /// `None` can be considered a StatusCode of `0`.
     #[ignore_heap_size_of = "Defined in hyper"]
@@ -134,6 +134,10 @@ impl Response {
             internal_response: None,
             return_internal: Cell::new(true)
         }
+    }
+
+    pub fn url(&self) -> Option<&Url> {
+        self.url.as_ref()
     }
 
     pub fn is_network_error(&self) -> bool {

--- a/components/net_traits/response.rs
+++ b/components/net_traits/response.rs
@@ -100,11 +100,11 @@ pub struct Response {
 }
 
 impl Response {
-    pub fn new() -> Response {
+    pub fn new(url: Url) -> Response {
         Response {
             response_type: ResponseType::Default,
             termination_reason: None,
-            url: None,
+            url: Some(url),
             url_list: RefCell::new(Vec::new()),
             status: Some(StatusCode::Ok),
             raw_status: Some((200, b"OK".to_vec())),

--- a/tests/unit/net/fetch.rs
+++ b/tests/unit/net/fetch.rs
@@ -383,8 +383,8 @@ fn test_fetch_response_is_opaque_filtered() {
     assert!(!fetch_response.is_network_error());
     assert_eq!(fetch_response.response_type, ResponseType::Opaque);
 
+    assert!(fetch_response.url().is_none());
     assert!(fetch_response.url_list.into_inner().len() == 0);
-    assert!(fetch_response.url.is_none());
     // this also asserts that status message is "the empty byte sequence"
     assert!(fetch_response.status.is_none());
     assert_eq!(fetch_response.headers, Headers::new());

